helper/schema/resource.go

```diff
@@ -187,8 +174,23 @@ type ExistsFunc func(*ResourceData, interface{}) (bool, error)
 type StateMigrateFunc func(
 	int, *terraform.InstanceState, interface{}) (*terraform.InstanceState, error)
 
+type StateUpgrader struct {
+	// Version is the version schema that this Upgrader will handle, converting
+	// it to Version+1.
+	Version int
+
+	// Type describes the schema that this function can upgrade. Type is
+	// required to decode the schema if the state was stored in a legacy
+	// flatmap format.
+	Type cty.Type
+
+	// Upgrade takes the JSON encoded state and the provider meta value, and
+	// upgrades the state one single schema version.
+	Upgrade StateUpgradeFunc
+}
+
 // See Resource documentation.
-type StateUpgradeFunc func(int, map[string]interface{}, interface{}) (map[string]interface{}, error)
+type StateUpgradeFunc func(map[string]interface{}, interface{}) (map[string]interface{}, error)
```

Comment:
I didn't catch this on the first pass I guess because I read things in a different order and so had it fresh in my mind, but I think it'd be helpful to name these arguments (`rawState` and `meta`, I guess?) just to make it a bit clearer what they are, since the types are too generic to guess from them alone.